### PR TITLE
Free tests & benchmarks

### PR DESCRIPTION
**Pull request by PhBastiani**
Purpose: evaluate the performance of a trampolined algorithm (calculation of Fibonacci sequences as support).
=> benchmarks Arrow / Cats for comparaison.

Trampolined version : ration Arrow / Cats > 4.

Note : i added benchmarks with no JVM warmup and with a single iteration : theses could be compared to memoized calculations; but, they rely on very imprecise results !

### CODE REVIEW

**Comment by rachelcarmena** [comment]
Thanks @PhBastiani  :tada: Please, I just miss its execution at the end:

* https://github.com/arrow-kt/arrow-incubator/blob/master/.github/workflows/build_arrow-incubator.yml

This file can be useful as an example:

* https://github.com/arrow-kt/arrow-fx/blob/master/.github/workflows/build_arrow-fx.yml

Thank you so much!

**Comment by PhBastiani** [comment]
Yes, I'll add that tomorrow morning.

@rachelcarmena: note that running benchmarks is a pain. The jmh plugins does not seem to release java resources properly.

Is there a way to display the bench reports ?

File: .github/workflows/build_arrow-incubator.yml [hunk 1]
```diff
@@ -26,3 +26,4 @@ jobs:
         git clone https://github.com/arrow-kt/arrow.git
     - name: Build with Gradle
       run: $BASEDIR/arrow/scripts/project-build.sh arrow-incubator
+      ./gradlew :arrow-benchmarks-free:jmhClasses
```
Comment (rachelcarmena) [inline .github/workflows/build_arrow-incubator.yml]:
```suggestion
      run: |
          $BASEDIR/arrow/scripts/project-build.sh arrow-incubator
          ./gradlew :arrow-benchmarks-free:jmhClasses
```

**Comment by rachelcarmena** [comment]
> Yes, I'll add that tomorrow morning.
> 
> @rachelcarmena: note that running benchmarks is a pain. The jmh plugins does not seem to release java resources properly.

Thanks @PhBastiani !! I'm reviewing it right now because it seems the report isn't being generated. And it wasn't being generated before the move... I'll let you know when I have news!

**Comment by rachelcarmena** [comment]
> Thanks @PhBastiani !! I'm reviewing it right now because it seems the report isn't being generated. And it wasn't being generated before the move... I'll let you know when I have news!

Ok... `:jmhClasses` task just compiles benchmark code. There are other tasks:
* `:jmh` task to run the benchmarks
* `:jmhReport` task to generate a HTML report from `build/reports/benchmarks.json`. That report could be provided to be downloaded from GitHub Actions.

**Comment by PhBastiani** [comment]
`:jmhClasses`  validates the user code
`:jmhReport` is not mentioned in the doc; and, it requires the json result !
`:jmh` is the global task... This task depends on `:jmhJar` but this one seem unsafe

**Comment by rachelcarmena** [comment]
> `:jmhClasses` validates the user code
> `:jmhReport` is not mentioned in the doc; and, it requires the json result !
> `:jmh` is the global task... This task depends on `:jmhJar` but this one seem unsafe

Right @PhBastiani :raised_hands: I wrote it a previous comment :+1: 

`jmhReport` tasks come from another plugin: https://github.com/jzillmann/gradle-jmh-report

**Comment by rachelcarmena** [comment]
@PhBastiani , I created an issue to compare results on pull requests: https://github.com/arrow-kt/arrow-fx/issues/76. Then, we can do the same for this case :raised_hands:

**Comment by rachelcarmena** [comment]
Sorry @PhBastiani , my mistake. Please, remove `tasks.jmh.finalizedBy tasks.jmhReport`.

File: .github/workflows/build_arrow-incubator.yml [hunk 2]
```diff
@@ -26,4 +26,15 @@ jobs:
...
+    - name: Run benchmarks
+      run: |
+        ./gradlew :arrow-benchmarks-free:jmhForArrow
+        ./gradlew :arrow-benchmarks-free:jmhForCats
+    - name: Run benchmark comparison
+      env:
+        CI_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      run: |
+        export PULL_REQUEST_NUMBER=$(echo $GITHUB_REF | cut -d/ -f3)
+        ./gradlew :arrow-benchmarks-free:compareBenchmarksCI
```
Comment (rachelcarmena) [inline .github/workflows/build_arrow-incubator.yml:40]:
```suggestion
        echo "ref: $GITHUB_REF"
        echo "sha: $GITHUB_SHA"
        export PULL_REQUEST_NUMBER=$(echo $GITHUB_REF | cut -d/ -f3)
        echo "pr-number: $PULL_REQUEST_NUMBER"
        ./gradlew :arrow-benchmarks-free:compareBenchmarksCI
```

I'd need these data to discover the reason of 403 error when creating the comment.

**Comment by rachelcarmena** [comment]
@PhBastiani :raised_hands: There is an issue with Hood and pull requests from a forked repository. I'm helping the team to solve it as soon as possible :+1:

**Comment by nomisRev** [comment]
@rachelcarmena what's the state of this?

**Comment by rachelcarmena** [comment]
> @rachelcarmena what's the state of this?

Sorry for the delay @nomisRev !! It's solved. However benchmarks aren't being used right now...

**Pull request closed without merging**